src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,252 @@
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.References
+{
+    [Export(typeof(IReferenceCleanupService))]
+    internal partial class ReferenceCleanupService : IProjectSystemReferenceCleanupService
+    {
+        ReferenceHandler _projectReferenceHandler = new ProjectReferenceHandler();
+        ReferenceHandler _packageReferenceHandler = new PackageReferenceHandler();
+        ReferenceHandler _assemblyReferenceHandler = new AssemblyReferenceHandler();
+        ReferenceHandler _sdkReferenceHandler = new SdkReferenceHandler();
+
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IVsUIService<DTE> _dte;
+        private readonly IVsUIService<SVsSolution, IVsSolution> _solution;
+
+        private Dictionary<ReferenceType, string> _referenceTypes = new Dictionary<ReferenceType, string>();
+
+        [ImportingConstructor]
+        public ReferenceCleanupService(ConfiguredProject configuredProject, IVsUIService<SDTE, DTE> dte, IVsUIService<SVsSolution, IVsSolution> solution)
+        {
+            _configuredProject = configuredProject;
+            _dte = dte;
+            _solution = solution;
+        }
+
+        public Task<string> GetProjectAssetsFilePathAsync(string projectPath, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        public async Task<ImmutableArray<ProjectSystemReferenceInfo>> GetProjectReferencesAsync(string projectPath, string targetFrameworkMoniker, CancellationToken cancellationToken)
+        {
+            List<ProjectSystemReferenceInfo> references;
+
+            try
+            {
+                var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath);
+
+                references = await GetAllReferencesInConfiguredProjectAsync(activeConfiguredProject);
+            }
+            catch
+            {
+                throw new InvalidProjectFileException();
+            }
+
+            return references.ToImmutableArray();
+        }
+
+        private async Task<ConfiguredProject> GetActiveConfiguredProjectByPathAsync(string projectPath)
+        {
+            var unconfiguredProjectPath = FindUnconfiguredProjectByPath(projectPath);
+
+            var activeConfiguredProject = await unconfiguredProjectPath.GetSuggestedConfiguredProjectAsync();
+            if (activeConfiguredProject is null)
+            {
+                throw new InvalidProjectFileException();
+            }
+
+            return activeConfiguredProject;
+        }
+
+        private UnconfiguredProject FindUnconfiguredProjectByPath(string projectPath)
+        {
+            var unconfigProjects = _configuredProject.Services.ProjectService.LoadedUnconfiguredProjects;
+
+            var unconfiguredProject = unconfigProjects.First(project =>
+                StringComparers.Paths.Equals((string)project.FullPath, projectPath));
```

Comment:
`First` throws if not found. Is that ok?

Reply:
Yes. I was thinking Roslyn should handle the case when the project is incorrect/typo/not found.